src/x86_64/x86_64.h

```diff
@@ -104,6 +104,11 @@ static inline void atomic_clear_bit(u64 *target, u64 bit)
     asm volatile("lock btrq %1, %0": "+m"(*target):"r"(bit) : "memory");
 }
 
+static inline void atomic_add(u64 *target, u64 num)
```

Comment:
The existing fetch_and_add() in x86.h takes 32-bit values, here we have 64-bit values. I renamed atomic_add() to fetch_and_add_64() and changed it to return the fetched value (even though the current callers to don't use the return value).

Reply:
The existing fetch_and_add takes a 'word' pointer, which should be 32-bit in stage2 and 64-bit in stage3. If it's 32-bit in stage3, then it's a bug.

But I don't think it is...note the QWORD in e.g.:

```
ffffffff80011f70:       48 89 73 20             mov    QWORD PTR [rbx+0x20],rsi                                                                                                                                                                
static inline word fetch_and_add(word *variable, word value)                                                                                                                                                                                   
{                                                                                                                                                                                                                                              
    asm volatile("lock; xadd %0, %1" : "+r" (value), "+m" (*variable) :: "memory", "cc");                                                                                                                                                      
ffffffff80011f74:       48 c7 c0 ff ff ff ff    mov    rax,0xffffffffffffffff                                                                                                                                                                  
ffffffff80011f7b:       f0 48 0f c1 43 08       lock xadd QWORD PTR [rbx+0x8],rax                                                                                                                                                              
                                                                                                                                                                                                                                               
    word n = fetch_and_add(&m->count, (word)-1);                                                                                                                                                                                               
    if (n == 1) {                                                                                                                                                                                                                              
ffffffff80011f81:       48 83 f8 01             cmp    rax,0x1                                                                                                                                                                                 
ffffffff80011f85:       74 09                   je     ffffffff80011f90 <merge_join+0x30>                                                                                                                                                      
        apply(m->completion, m->last_status);                                                                                                                                                                                                  
        deallocate_closure(m->apply);                                                                                                                                                                                                          
        deallocate(m->h, m, sizeof(struct merge));                                                                                                                                                                                             
        closure_finish();                                                                                                                                                                                                                      
    }                                                                                                                                                                                                                                          
}                                                                                                                                                                                                                                              
```

Maybe the new one is a little better, though, because the compiler seems to generate an xadd only if the result is being used, and otherwise just an add. If the result is used, the disassembly is identical to the original.

Anyhow, it seems unnecessary to have two variants here.

Reply:
You are right, I saw the definition of word in def32.h but didn't notice that there is another definition in x86_64.h.